### PR TITLE
Synchornize etcd cluster in registrator on service registration

### PR DESCRIPTION
**Pull request by hawka**

### CODE REVIEW

**Comment by joshk0** [comment]
Would love some details of how you tested this. I have no idea how this code is going to play out, as i'm not familiar enough with the internals of registrator.

My suggestion for testing would be to start up registrator on bastion (e.g. having connectivity to our real etcd cluster) but pass a different prefix, like /amalia. Then start and stop some containers on bastion and see what gets written there, and if there is any evidence it is using the new URLs.

**Comment by hawka** [comment]
I haven't, yet -- the base project has no existing tests and it's tough to unit test without going out of scope. My plan was to set up something on bastion, and what you suggest sounds workable!

**Comment by hawka** [comment]
Discussed offline. Logging used for testing:
```
diff --git a/etcd/etcd.go b/etcd/etcd.go
index f1031c3..998fac1 100644
--- a/etcd/etcd.go
+++ b/etcd/etcd.go
@@ -77,13 +77,21 @@ func (r *EtcdAdapter) Ping() error {
 
 func (r *EtcdAdapter) syncEtcdCluster() error {
        var result bool
+       var newTopo []string
        if r.client != nil {
                result = r.client.SyncCluster()
+               if result {
+                       newTopo = r.client.GetCluster()
+               }
        } else {
                result = r.client2.SyncCluster()
+               if result {
+                       newTopo = r.client2.GetCluster()
+               }
        }
 
        if result {
+               log.Printf("AMALIA ::: %v", newTopo)
                return nil
        }
        return errors.New("sync cluster was unsuccessful")
```

**Comment by hawka** [comment]
Tested locally with 3 etcds, seems to be working.

**Comment by joshk0** [comment]
LGTM, *but*...

One wrinkle is that we might want to deploy this to production prior to the registrator guys a) accepting the change, b) cutting the release, c) updating the published container. This means we need a way to build the docker image from your fork, and publish it somewhere we can get at it.

My suggestion is that you:
- add the pippio/registrator repo as a submodule pinned to your commit
- add the build of the registrator Dockerfile to devops/ci.sh
- make it possible to publish a hacked registrator container to our private-registry
- change devops/units/registrator.service to use the hacked registrator image.

I am happy with this change as the PR to registrator upstream, then I would do a follow up for the ops piece in our repo.

**Comment by joshk0** [comment]
I also want to mention that in our meatspace discussion, I brought up the idea of logging when the cluster topology changes as a result of a call to SyncCluster. This can help with debugging any teething issues related to this change. However, it might be overkill because there's not much logging in registrator otherwise except for when services are registered, deregistered or refreshed. I would personally like it for our own use, but it doesn't necessarily make sense upstream. Leaving that as an open discussion.

File: etcd/etcd.go [hunk 1]
```diff
@@ -81,8 +103,16 @@ func (r *EtcdAdapter) Register(service *bridge.Service) error {
 
 	if err != nil {
 		log.Println("etcd: failed to register service:", err)
+		return err
 	}
-	return err
+
+	err = r.syncEtcdCluster()
```
Comment (jgraettinger) [inline etcd/etcd.go]: If we're doing this each Register() anyway, should it happen before the attempted Set()?

Comment (joshk0) [inline etcd/etcd.go]: Good catch, not used to the diff context in github anymore. All should occur before performing any real operation with the client.

**Comment by hawka** [comment]
Updated w.r.t. comments

**Comment by joshk0** [comment]
LGTM for creating the upstream PR, looking forward to the devops piece.

**Pull request merged**